traitsui/wx/image_enum_editor.py

```diff
@@ -113,13 +117,24 @@ def init(self, parent):
         """ Finishes initializing the editor by creating the underlying toolkit
             widget.
         """
-        self._create_image_grid(parent)
+        super(CustomEditor, self).init(parent)
+
+        # Create the panel to hold the ImageControl buttons:
+        self.control = TraitsUIPanel(parent, -1)
+        self._create_image_grid()
+
+    def rebuild_editor(self):
+        self._create_image_grid()
 
-    def _create_image_grid(self, parent):
+    def _create_image_grid(self):
         """ Populates a specified window with a grid of image buttons.
         """
-        # Create the panel to hold the ImageControl buttons:
-        self.control = panel = TraitsUIPanel(parent, -1)
+        # Clear any existing content:
+        panel = self.control
+        panel.SetSizer(None)
+        for child in panel.GetChildren():
+            _popEventHandlers(child)
+        panel.DestroyChildren()
```

Comment:
Why is this needed? This is worrying because this could be blindly removing components depended on by other change handlers. This editor should know exactly what content are in the control.

Reply:
This is needed to remove previous buttons, otherwise they start overlapping. In this case the only children are the buttons that are re-added just below.  The same approach is used in `Enum` and `CheckList` custom editors.

Reply:
I see. I think I read this with `init` and was surprised.  Could you confirm that this block is only required by `rebuild_editor` only?  If so, it would make more sense if this block is moved to `rebuild_editor` instead.

I could not find the same use of `_popEventHandlers` in the other editors. `_popEventHandlers` is private to `traitsui.wx.toolkit` and it only recognizes a subtype of event handler that is created there. I am still not convinced the connected components are disconnected. Looks like `ImageControl` needs to provide a method for `Unbind`-ing the events so that this container can call it to explicitly disconnect event handlers.

Reply:
It's not necessary in the `init` - there are no children at that time and the operation essentially does nothing.

The other uses suffer from https://github.com/enthought/traitsui/issues/835 and calling `_popEventHandlers` seems to fix it. I assume `wx` code complains unless all handlers are removed, and since `_popEventHandlers` removes the error, I assume that it removes all attached handlers.

Re privacy of `_popEventHandlers`, it's not necessary to call it directly, there is a toolkit method `destroy_children` that does what is needed here. But to get it I would either have to instantiate `traitsui.wx.toolkit.GUIToolkit` or go through `traitsui.toolkit` to get the already instantiated one. Using `_popEventHandlers` directly seemed most appropriate for the simple use case.

Reply:
Thank you. Let's move this block to `rebuild_editor` and let's call `GUIToolkit().destroy_children()`. Public interface is always preferred to private ones. If the logic of `destroy_children` needs to be changed, there is only one place to make such changes. We can worry about optimization later.

I still have reservation with this `destroy_children` in these `rebuild_editor`, and probably you do too. 

This is what I arrived at by reading the code, but I haven't tested this:

The `destroy_children` (or `_popEventHandlers`) probably does remove some lingering event handlers that are "pushed" recursively when the the UI is first initialized:
https://github.com/enthought/traitsui/blob/64858b8c9ce9189a8c9cfde60610e4bd12f30962/traitsui/ui.py#L406-L409
This may explain why #842 can be resolved by calling `_popEventHandlers`.

For these editors with `rebuild_editors, if `values` are modified many times, the first call to `_popEventHandlers` remove handlers that are not put back by `_create_image_grid `. In other words, the N-th (N>1) call to the `_popEventHandlers`  will find no event handlers to be removed.

`ImageControl` still binds event handlers and provides no way to unbind them.  However the error in #842 suggests that it only cares about the event handlers added by "pushing", so the error may not be relevant when the `values` are modified a second time.  I guess we can test that quite easily?

(Though we should still make sure what put up is taken down even if it does not cause errors now... that is perhaps another issue for `ImageControl` to provide a dispose method.)

Reply:
Moved `delete_children` to `rebuild_editor`. I don't think it's a problem that `_popEventHandlers` won't find any handlers to remove after the first time, it will be a simple no-op.

I opened #856 to add cleanup to `ImageControl`